Libraries/PeriphDrivers/Source/I3C/i3c_reva.c

```diff
@@ -0,0 +1,941 @@
+} mxc_i3c_reva_ibiresp_t;
+
+#define MXC_S_I3C_REVA_TARG_CAP1_CCCH_BASIC (0x1 << MXC_F_I3C_REVA_TARG_CAP1_CCCH_POS)
+#define MXC_S_I3C_REVA_TARG_CAP1_CCCH_LIMITS (0x2 << MXC_F_I3C_REVA_TARG_CAP1_CCCH_POS)
+#define MXC_S_I3C_REVA_TARG_CAP1_IBI_EVENTS_IBI (0x1 << MXC_F_I3C_REVA_TARG_CAP1_IBI_EVENTS_POS)
+#define MXC_S_I3C_REVA_TARG_CAP1_IBI_EVENTS_PAYLOAD (0x2 << MXC_F_I3C_REVA_TARG_CAP1_IBI_EVENTS_POS)
+#define MXC_S_I3C_REVA_TARG_CAP1_IBI_EVENTS_HOTJOIN (0x4 << MXC_F_I3C_REVA_TARG_CAP1_IBI_EVENTS_POS)
+
+#define MXC_I3C_REVA_ADDR_INVALID 0x00U
+
+/* **** Globals **** */
+/**
+ * @brief I3C controller instances.
+ *
+ */
+mxc_i3c_reva_controller_t controller[MXC_CFG_I3C_INSTANCES];
+
+/* **** Functions **** */
+int MXC_I3C_RevA_Init(mxc_i3c_reva_regs_t *i3c, bool targetMode, uint8_t staticAddr)
+{
+    int ret = E_NO_ERROR;
+    int idx;
+    uint8_t val;
+
+    idx = MXC_I3C_GET_IDX((mxc_i3c_regs_t *)i3c);
+    if (idx < 0) {
+        return E_BAD_PARAM;
+    }
+
+    if (!targetMode) {
+        /* Controller mode initialization */
+
+        /* SCL-to-SDA skew */
+        ret = MXC_I3C_RevA_SetSkew(i3c, 0);
+        if (ret < 0) {
+            return ret;
+        }
+
+        /* High-Keeper implementation selection */
+        ret = MXC_I3C_RevA_SetHighKeeperMode(i3c, MXC_I3C_HIGH_KEEPER_OFF);
+        if (ret < 0) {
+            return ret;
+        }
+
+        /* 2. If using fifos, set tx rx fifo trigger levels */
+        if (i3c->targ_cap1 & MXC_F_I3C_REVA_TARG_CAP1_RXFIFO_CFG) {
+            MXC_I3C_RevA_SetRXTXThreshold(i3c, MXC_I3C_REVA_RX_TH_NOT_EMPTY,
+                                          MXC_I3C_REVA_TX_TH_ALMOST_FULL);
+        }
+
+        /* 3. Optionally write IBIRULES reg to optimize response to incoming IBIs */
+
+        /* 4. Enable controller mode */
+        SET_FIELD(cont_ctrl0, CONT_CTRL0_EN, MXC_V_I3C_REVA_CONT_CTRL0_EN_ON);
+    } else {
+        /* Target mode initialization */
+        /* 1. After reset, write setup registers needed for optional features */
+
+        if (i3c->targ_cap1 & MXC_F_I3C_REVA_TARG_CAP1_RXFIFO_CFG) {
+            MXC_I3C_RevA_SetRXTXThreshold(i3c, MXC_I3C_REVA_RX_TH_NOT_EMPTY,
+                                          MXC_I3C_REVA_TX_TH_ALMOST_FULL);
+        }
+
+        /* Set BCR, DCR */
+        val = 0 << 6; /* Device role: target */
+        if (i3c->targ_cap1 & MXC_S_I3C_REVA_TARG_CAP1_CCCH_LIMITS) {
+            val |= 1 << 0; /* Max data speed limitation */
+        }
+        if (i3c->targ_cap1 & MXC_S_I3C_REVA_TARG_CAP1_CCCH_BASIC) {
+            val |= 1 << 5; /* Supports advanced capabiliries */
+        }
+        if (i3c->targ_cap1 & MXC_S_I3C_REVA_TARG_CAP1_IBI_EVENTS_IBI) {
+            val |= 1 << 1; /* Supports IBI generation */
+        }
+        if (i3c->targ_cap1 & MXC_S_I3C_REVA_TARG_CAP1_IBI_EVENTS_PAYLOAD) {
+            val |= 1 << 2; /* MDB and additional data bytes may follow the IBI */
+        }
+        val |= 1 << 3; /* Offline capable */
```

Comment:
Might be a good idea to bring these out as enumerated definitions in the register file so you don't have to hardcode these values.